cmd/k8sagent/main_nix.go

```diff
@@ -0,0 +1,198 @@
+
+https://jujucharms.com/
+
+The k8sagent command can also forward invocations over RPC for execution by the
+juju unit agent. When used in this way, it expects to be called via a symlink
+named for the desired remote command, and expects JUJU_AGENT_SOCKET_ADDRESS and
+JUJU_CONTEXT_ID be set in its model.
+`
+
+const (
+	// exit_err is the value that is returned when the user has run juju in an invalid way.
+	exit_err = 2
+	// exit_panic is the value that is returned when we exit due to an unhandled panic.
+	exit_panic = 3
+)
+
+type k8sAgentLogWriter struct {
+	target io.Writer
+}
+
+func (w *k8sAgentLogWriter) Write(entry loggo.Entry) {
+	if strings.HasPrefix(entry.Module, "unit.") {
+		fmt.Fprintln(w.target, w.unitFormat(entry))
+	} else {
+		fmt.Fprintln(w.target, loggo.DefaultFormatter(entry))
+	}
+}
+
+func (w *k8sAgentLogWriter) unitFormat(entry loggo.Entry) string {
+	ts := entry.Timestamp.In(time.UTC).Format("2006-01-02 15:04:05")
+	// Just show the last element of the module.
+	lastDot := strings.LastIndex(entry.Module, ".")
+	module := entry.Module[lastDot+1:]
+	return fmt.Sprintf("%s %s %s %s", ts, entry.Level, module, entry.Message)
+}
+
+func k8sAgentCommand(ctx *cmd.Context) (cmd.Command, error) {
+	// Assuming an average of 200 bytes per log message, use up to
+	// 200MB for the log buffer.
+	bufferedLogger, err := logsender.InstallBufferedLogWriter(loggo.DefaultContext(), 1048576)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Set the default transport to use the in-process proxy
+	// configuration.
+	if err := proxy.DefaultConfig.Set(proxyutils.DetectProxies()); err != nil {
+		return nil, errors.Trace(err)
+	}
+	if err := proxy.DefaultConfig.InstallInDefaultTransport(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	k8sAgent := jujucmd.NewSuperCommand(cmd.SuperCommandParams{
+		Name: "k8sAgent",
+		Doc:  k8sAgentDoc,
+	})
+	k8sAgent.Log.NewWriter = func(target io.Writer) loggo.Writer {
+		return &k8sAgentLogWriter{target: target}
+	}
+
+	k8sAgent.Register(initcommand.New())
+	k8sAgent.Register(unitcommand.New(ctx, bufferedLogger))
+	return k8sAgent, nil
+}
+
+func mainWrapper(f commandFactotry, args []string) (code int) {
+	ctx, err := cmd.DefaultContext()
+	if err != nil {
+		cmd.WriteError(os.Stderr, err)
+		os.Exit(exit_err)
+	}
+	switch filepath.Base(args[0]) {
+	case names.K8sAgent:
+		code = f.k8sAgentCmd(ctx, args)
+	case names.JujuRun:
+		code = f.jujuRun(ctx, args)
+	case names.JujuDumpLogs:
+		code = f.jujuDumpLogs(ctx, args)
+	case names.JujuIntrospect:
+		code = f.jujuIntrospect(ctx, args)
+	default:
+		// This should never happen unless jujuc was missing and hooktools were misconfigured.
+		err = errors.New("k8sagent always expects to use jujuc for hook tools")
+		cmd.WriteError(ctx.Stderr, err)
+		os.Exit(exit_err)
+	}
+	return code
+}
+
+// CAAS implementation currently only works for Linux.
+func main() {
+	defer func() {
+		if r := recover(); r != nil {
+			buf := make([]byte, 4096)
```

Comment:
I do wonder whether or not 4096 is big enough for a big stack trace.

Reply:
This is the current jujud config, I assume k8sagent would need the size less than the  controller needs. 
And we set the all=False, so we only have the current gorounte's trace.